api/maven-api-core/src/main/java/org/apache/maven/api/DependencyProperties.java

```diff
@@ -33,27 +36,226 @@
+        public Key<V> intern() {
+            Key<?> previous;
+            synchronized (INTERNS) {
+                previous = INTERNS.putIfAbsent(name, this);
+            }
+            if (previous == null) {
+                return this;
+            }
+            if (equals(previous)) {
+                return (Key<V>) previous;
+            }
+            throw new IllegalStateException("Key " + name + " already exists for a different class of values.");
+        }
+
+        /**
+         * {@return a string representation of this key}.
+         * By default, this is the name of this key.
+         */
+        @Override
+        public String toString() {
+            return name;
+        }
+
+        /**
+         * {@return an hash code value for this key}.
+         */
+        @Override
+        public int hashCode() {
+            return 7 + name.hashCode() + 31 * valueType.hashCode();
+        }
+
+        /**
+         * Compares this key with the given object for equality.
+         * Two keys are considered equal if they have the same name
+         * and are associated to values of the same class.
+         *
+         * @param obj the object to compare with this key
+         * @return whether the given object is equal to this key
+         */
+        @Override
+        public boolean equals(Object obj) {
+            if (this == obj) {
+                return true;
+            }
+            if (obj != null && getClass() == obj.getClass()) {
+                final Key<?> other = (Key<?>) obj;
+                return name.equals(other.name) && valueType.equals(other.valueType);
+            }
+            return false;
+        }
+    }
+
+    /**
+     * The dependency type. The {@linkplain Key#name() name} of this property
+     * is equal to the {@code ArtifactProperties.TYPE} value.
+     */
+    Key<String> TYPE = new Key<>("type", String.class).intern();
+
+    /**
+     * The dependency language. The {@linkplain Key#name() name} of this property
+     * is equal to the {@code ArtifactProperties.LANGUAGE} value.
+     */
+    Key<String> LANGUAGE = new Key<>("language", String.class).intern();
+
+    /**
+     * Types of path (class-path, module-path, …) where the dependency can be placed.
+     * For most deterministic builds, the array length should be 1. In such case,
+     * the dependency will be unconditionally placed on the specified type of path
+     * and no heuristic rule will be involved.
+     *
+     * <p>It is nevertheless common to specify two or more types of path. For example,
+     * a Java library may be compatible with either the class-path or the module-path,
+     * and the user may have provided no instruction about which type to use. In such
+     * case, the plugin may apply rules for choosing a path. See for example
+     * {@link JavaPathType#CLASSES} and {@link JavaPathType#MODULES}.</p>
+     */
+    Key<PathType[]> PATH_TYPES = new Key<>("pathTypes", PathType[].class).intern();
```

Comment:
`Collection<PathType>` so that we can ensure immutability ?

Reply:
The use of collection is not fully type-safe with the `Key<V>` mechanism, because of generic type erasure. A `java.lang.Class` can represent `Collection`, but cannot represent `Collection<PathType>`. By contrast, the use of array is fully type-safe (can be represented by a `Class`), and the immutability may come in the future with [Frozen Arrays](https://openjdk.org/jeps/8261007).

A simpler alternative would be to not manage `PathType` in `DependencyProperties`, but instead as an ordinary method directly in `Dependency`. Use of collections would be type-safe there.

Reply:
What about using `java.lang.reflect.Type` instead of `Class<?>` here and in the value converters ?

Reply:
I think it would not avoid the "unchecked cast" compiler warning. An easier approach would be to replace this property by an ordinary `Set<PathType> getPathTypes()` method in the `Dependency` interface.

Reply:
Replaced `PathType[]` by `Set<PathType>`. It introduces a compiler warning (disabled with `@SuppressWarnings`), but only in a code that we control. So hopefully, this is okay.